### ISSUE TITLE
New rule proposal: compact-accessor

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12277]
**Please describe what the rule should do:**

If an accessor property has both getter and setter functions, this rule requires their definitions to be one after another in object literals/class definitions.

Optionally, the rule can require their order (get before set, or set before get).

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style (layout)

**Provide 2-3 code examples that this rule will warn about:**

```js
/* eslint compact-accessor:error*/
var foo = {
    get a(){},
    bar: 5,
    set a(val){}
};

var foo = {
    get a(){},
    ...bar,
    set a(val){}
}

class foo {
    get a(){}
    bar(){}
    set a(val){}
}
```

```js
/* eslint compact-accessor:["error", "getBeforeSet"]*/
var foo = {
    set a(val){},
    get a(){}
}

class foo {
    set a(val){}
    get a(){}
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

While it is allowed, I don't think there is a real use case for non-consecutive definitions of accessor functions for the same property, and it should be a best practice to avoid this feature. Also, the additional option can enforce a consistent style (e.g., `get` before `set`).

I'm not sure about the rule's name, there could be a better word than `compact` for this.

Notes:

* The rule does not enforce the existence of the pair for a getter/setter (that's `accessor-pairs`). 
* The rule does not require grouping of all accessor properties at the same place, it works with each individually.
* If a property has a duplicate getter or setter, such property will be ignored by this rule and caught by other rules (`no-dupe-keys` and `no-dupe-class-members`)

Problem: This rule is overlapping with `sort-keys`. But, one might want to enforce this without sorting. Also, `sort-keys` doesn't work with classes and doesn't enforce get/set order.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by ilyavolodin** [comment]
I think it's a reasonable rule to add. But would need a different name, compact-accessors doesn't really describe it well. Maybe paired-accessors?

**Comment by kaicataldo** [comment]
I like `paired-accessors`!

**Comment by ljharb** [comment]
They're already paired because they have the same property name; this seems more like `grouped-accessors`.

**Comment by mdjermanovic** [comment]
I don't like `compact-accessors` name much, too. 

`paired-accessors` looks good, though it might imply enforcing the existence of the pair? Also, there is already the `accessor-pairs` rule, so two similar names could be confusing?

`grouped-accessors` also looks good, but it could be interpreted as enforcing all accessor properties (different names) to be defined at the same place?

Maybe `connected-accessor-functions`?

**Comment by ljharb** [comment]
grouped-accessor-pairs?

**Comment by mdjermanovic** [comment]
> grouped-accessor-pairs?

This sounds to me like a very good name for this purpose. If its apparent similarity with the existing name accessor-pairs isn't a problem.

**Comment by mdjermanovic** [comment]
I'll work on this over the weekend, should be ready for review soon.

**Comment by ljharb** [comment]
#12331
